package-name-map/src/test/package-name-map_test.ts

```diff
@@ -0,0 +1,226 @@
+/**
+ * Copyright 2018 Google Inc. All Rights Reserved.
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import chai = require('chai');
+
+import {PackageNameMap, isPathSegmentPrefix} from '../package-name-map.js';
+
+const {assert} = chai;
+
+suite('PackageNameMap', () => {
+  suite('resolve', () => {
+    const referrerURL = 'http://foo.com/';
+
+    suite('does not modify already valid specifiers', () => {
+      const map = new PackageNameMap({
+        packages: {
+          foo: {
+            main: 'index.js',
+          },
+        },
+      });
+
+      test('does not modify URLs', () => {
+        assert.equal(
+          map.resolve('https://bar.com', referrerURL),
+          'https://bar.com'
+        );
+      });
+
+      test('does not modify valid paths', () => {
+        assert.equal(map.resolve('/foo', referrerURL), '/foo');
+        assert.equal(map.resolve('./foo', referrerURL), './foo');
+        assert.equal(map.resolve('../foo', referrerURL), '../foo');
+      });
+    });
+
+    suite('top-level package names', () => {
+      const map = new PackageNameMap({
+        packages: {
+          app: {
+            main: 'src/index.js',
```

Comment:
This is interesting because consider `referrerURL` of `https://foo.com/js/entry.js`. What should the result be, since you are using a relative URL? Should the answer be different for `https://foo.com/js/nested/entry.js`? Should the answer depend on where the package name map itself is hosted?

It seems like either we need to disallow relative paths, or we need to turn them into absolute paths ASAP, probably using the package name map's URL (which is distinct from the referrerURL).

Reply:
does it need to be the package-name-map's URL, or the document's baseURL?

Reply:
Consider `https://example.com/a.html` with `<script type="packagemap" src="b/map.json">` which contains `pkg: { main: "c.js" }`.

Should `import "pkg"` give `https://example.com/b/c.js` or `https://example.com/c.js`? I'm not sure, but I lean toward the former; when I see `c.js` inside `https://example.com/b/map.json` I'd probably expect that to point to `https://example.com/b/c.js`.

Reply:
I made the PackageNameMap constructor take a baseURI - it's agnostic as to where it comes from.